dex/encrypt/encrypt.go

```diff
@@ -0,0 +1,247 @@
+	"golang.org/x/crypto/poly1305"
+)
+
+// Crypter is an interface for an encryption key and encryption/decryption
+// algorithms. Create a Crypter with the NewCrypter function.
+type Crypter interface {
+	// Encrypt encrypts the plaintext.
+	Encrypt(b []byte) ([]byte, error)
+	// Decrypt decrypts the ciphertext created by Encrypt.
+	Decrypt(b []byte) ([]byte, error)
+	// Serialize serializes the Crypter. Use the Deserialize function to create
+	// a Crypter from the resulting bytes. Deserializing requires the password
+	// used to create the Crypter.
+	Serialize() []byte
+	// Close zeros the encryption key. The Crypter is useless after closing.
+	Close()
+}
+
+const (
+	// defaultTime is the default time parameter for argon2id key derivation.
+	defaultTime = 1
+	// defaultMem is the default memory parameter for argon2id key derivation.
+	defaultMem = 64 * 1024
+	// KeySize is the size of the encryption key.
+	KeySize = 32
+	// SaltSize is the size of the argon2id salt.
+	SaltSize = 16
+)
+
+// Big-endian encoder.
+var intCoder = encode.IntCoder
+
+// Key is 32 bytes.
+type Key [KeySize]byte
+
+// Salt is randomness used as part of key deriviation. This is different from
+// the salt generated during xchacha20poly1305 encryption, which is shorter.
+type Salt [SaltSize]byte
+
+// newSalt is the constructor for a salt based on randomness from crypto/rand.
+func newSalt() Salt {
+	var s Salt
+	_, err := rand.Read(s[:])
+	if err != nil {
+		panic("newSalt: " + err.Error())
+	}
+	return s
+}
+
+// argonParams is a set of parameters for key derivation.
+type argonParams struct {
+	time    uint32
+	memory  uint32
+	threads uint8
+}
+
+// NewCrypter derives an encryption key from a password string.
+func NewCrypter(pw string) Crypter {
+	return newArgonPolyCrypter(pw)
+}
+
+// Deserialize deserializes the Crypter for the password.
+func Deserialize(pw string, encCrypter []byte) (Crypter, error) {
+	ver, pushes, err := encode.DecodeBlob(encCrypter)
+	if err != nil {
+		return nil, err
+	}
+	switch ver {
+	case 0:
+		return decodeArgonPoly_v0(pw, pushes)
+	default:
+		return nil, fmt.Errorf("unknown Crypter version %d", ver)
+	}
+}
+
+// argonPolyCryper is an encryption algorithm based on argon2id for key
+// derivation and xchacha20poly1305 for symmetric encryption.
+type argonPolyCrypter struct {
+	key    Key
+	tag    [poly1305.TagSize]byte
+	salt   Salt
+	params *argonParams
+}
+
+// newArgonPolyCrypter is the constructor for an argonPolyCrypter.
+func newArgonPolyCrypter(pw string) *argonPolyCrypter {
+	pwB := []byte(pw)
+	salt := newSalt()
+	threads := uint8(runtime.NumCPU())
+
+	keyB := argon2.IDKey(pwB, salt[:], defaultTime, defaultMem, threads, KeySize*2)
+	// The argon2id key is split into two keys, The encryption key is the first 32
+	// bytes.
+	encKeyB := keyB[:KeySize]
+	var encKey Key
+	copy(encKey[:], encKeyB)
+	// MAC key is the second 32 bytes.
+	polyKeyB := keyB[KeySize:]
+	var polyKey [KeySize]byte
+	copy(polyKey[:], polyKeyB)
```

Comment:
Same with `polyKeyB` unless you think it serves to document intent, but I think the LHS of `copy` does that.